### ISSUE TITLE
Make `CancellationTokenSource.Token` accessible after dispose

### ISSUE THREAD

**Issue by JamesNK** [dotnet/runtime#105811]
`CancellationTokenSource` guidance recommends disposing of the CTS once it is no longer used. However, after disposal the `Token` property throws `ObjectDisposedException` when accessed:

https://github.com/dotnet/runtime/blob/74c608d67d64675ff840c5888368669777c8aa2c/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L77-L87

I've seen this be a problem a number of times. For example, in ASP.NET Core a request can be aborted with a signal from the client. This would cancel the CTS that drives the `HttpContext.RequestAborted` cancellation token and it was then disposed. But server code could still be executing and access `HttpContext.RequestAborted`. Instead of returning the cancelled token, the code would get an ODE. The fix in ASP.NET Core was to not dispose a canceled CTS. We solved the problem, but it felt like a pit of failure that is easy to fall into.

Because of this situation, people who use CTS either need to:

* Be careful about not accessing the Token property after dispose (not an easy feat if dispose and API usage happen on different threads)
* Assign the token value to a field or property on the class (see below)
* Not dispose the CTS

Example of making a type safe by setting the cancellation token to a field:

```cs
public class Worker : IAsyncDisposable
{
    private readonly CancellationTokenSource _cts = new();
    private readonly CancellationToken _cancellationToken = _cts.Token; // field can be safely accessed
    private Task? _workerTask;

    public Worker()
    {
        _workerTask = Task.Run(() =>
        {
            while (_cancellationToken.IsCancellationRequested)
            {
                // do a thing
                await Task.Delay(1000, _cancellationToken);
            }
        });
    }

    public async ValueTask DisposeAsync()
    {
        _cts.Cancel();
        _cts.Dispose();
        try { await _workerTask; } catch (OperationCanceledException) { }
    }
}
```

This feels like it would be simpler and less of a pit of failure if `CancellationTokenSource.Token` could still be accessed after dispose.

**Comment by JamesNK** [comment]
@vladd Yes, that would fix this case. Every example I give there will always be a way to change the code to make it work. CTS isn't broken. But it could be made easier to use with one simple change.

How about this Blazor dialog component that has a CancellationTokenSource:

https://github.com/dotnet/aspire/blob/a41bd1baf4367f6b9394575d75dd995340b4795b/src/Aspire.Dashboard/Components/Dialogs/ExemplarsDialog.razor.cs

`OnViewDetailsAsync` is triggered by someone clicking on a button in the UI. It passes a cancellation token to a method that does some async work.

`Dispose` is triggered when the dialog is clicked. It cancels the CTS.

Now, I don't _think_ these two events can race because Blazor has the concept of the UI thread, but perhaps there is an exceptional situation where UI events come in out of order and they could, and it just hasn't happened yet. If `Dispose` is called and it's then followed by `OnViewDetailsAsync` then an `ObjectDisposedException` is thrown when accessing the `CancellationTokenSource.Token`.

**Comment by vladd** [comment]
@JamesNK I usually avoid patterns [like](https://github.com/dotnet/aspire/blob/a41bd1baf4367f6b9394575d75dd995340b4795b/src/Aspire.Dashboard/Components/Dialogs/ExemplarsDialog.razor.cs#L82)

```csharp
cts.Cancel();
cts.Dispose();
```

by the very same reason: `Cancel` is a signal to whoever is using the token, so I tend to maintain an inner list of those and await all of them to finish before disposing. (But I understand that this way is tedious and maybe an overkill.)

By the way, CancellationToken after disposal of CTS allows using [everything except WaitHandle](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMA6AFQAswBTAQzgEsoBzAbjzwGF6oBjWgBsh9YEwD2UChIDWtKAGUJEMIKL9gAZyIBeIlFoB3ABQBKTrl4Dho8VJnyoG4HpdbKchZbyaP1wSFzSz8yABEmLQAHCS1aYK5cTTIAJVpWSOBaMBNzPQA+EkwAThMAIn4+QKFaODKzCyIAeiaiIwkwWS08ADd6MCIqPjgat2SAdXomYAAJYZr2ZtbgGgkjHQB5MIBRIA) (which is a bit inconsistent, too).

**Comment by omghb** [comment]
I'm also having issues because of this behavior. We are using `CreateLinkedTokenSource` so dispose is important. We cannot control that a `token` might be cancelled right after dispose and so we get sometimes `System.ObjectDisposedException: CancellationTokenSource_Disposed`.

On StackOverflow a lot people are discussing this issue as well. An interesting post is here suggesting to not use `Dispose` and use `Cancel` instead. But for linked token source this might not work.
https://stackoverflow.com/a/51220106/28603541

**Workaround**
```C#
internal static class CancellationTokenSourceHelper
{
    public static void TryCancel(this CancellationTokenSource source)
    {
        try { source.Cancel(); }
        catch (ObjectDisposedException) { }
    }
}
```
